packages/backend/server/src/plugins/payment/service.ts

```diff
@@ -607,6 +633,72 @@ export class SubscriptionService {
     });
   }
 
+  async saveOnetimePaymentSubscription(user: User, invoice: UserInvoice) {
+    const savedSubscription = await this.db.userSubscription.findUnique({
+      where: {
+        userId_plan: {
+          userId: user.id,
+          plan: invoice.plan,
+        },
+      },
+    });
+
+    // TODO(@forehalo): time helper
+    const subscriptionTime =
+      (invoice.recurring === SubscriptionRecurring.Monthly ? 30 : 365) *
+      24 *
+      60 *
+      60 *
+      1000;
+
+    // extends the subscription time if exists
+    if (savedSubscription) {
+      if (!savedSubscription.end) {
+        throw new InternalServerError(
+          'Unexpected onetime subscription with no end date'
+        );
+      }
+
+      const period =
+        // expired, reset the period
+        savedSubscription.end <= new Date()
+          ? {
+              start: new Date(),
```

Comment:
should we extend the expiration date of the current subscription?

Reply:
what do you mean?

Reply:
i misunderstood the condition, there's nothing wrong with the logic